torch/ao/quantization/fx/convert.py

```diff
@@ -165,41 +166,117 @@ def _replace_observer_with_quantize_dequantize_node(
         # TODO(future PR): switch compute_dtype to is_dynamic
         # dynamic quantization
         node_type = "call_function"
+        is_dynamic = True
         if is_decomposed:
-            raise NotImplementedError("decomposed quantize_per_tensor_dynamic op not implemented yet")
+            dtype_ = to_underlying_dtype(dtype)
+            assert dtype_ in [torch.uint8, torch.int8], \
+                "only uint8 and int8 are supported in reference flow for " \
+                "dynamic quantization right now"
+            quant_min = activation_post_process.quant_min  # type: ignore[attr-defined]
+            quant_max = activation_post_process.quant_max  # type: ignore[attr-defined]
+            # note: scale and zero_point are missing for quantize_per_tensor op
+            # we'll need to get this from choose_qparams op, which we'll add after
+            # this stage
+            qparams = {
+                "_quant_min_": quant_min,
+                "_quant_max_": quant_max,
+                "_dtype_": dtype_
+            }
+            quantize_op = torch.ops.quantized_decomposed.quantize_per_tensor.tensor
         else:
             quantize_op = torch.quantize_per_tensor_dynamic
-        # TODO: get reduce range from observer
-        # reduce_range = activation_post_process.reduce_range
-        reduce_range = torch.backends.quantized.engine in ("fbgemm", "x86")
-        qparams = {"_dtype_": compute_dtype, "_reduce_range_": reduce_range}
+            # TODO: get reduce range from observer
+            # reduce_range = activation_post_process.reduce_range
+            reduce_range = torch.backends.quantized.engine in ("fbgemm", "x86")
+            qparams = {"_dtype_": compute_dtype, "_reduce_range_": reduce_range}
     elif dtype == torch.float16:
         node_type = "call_method"
         quantize_op = "to"
         qparams = {"_dtype_": dtype}
 
-    # 2. replace observer node with quant - dequant node
+    # 2. insert choose_qparams op and update the qparams list
+
+    # only needed for is_dynamic because the quantization parameters for static
+    # quantization is fixed (based on calibration), we don't need to compute it
+    # dynamically in the graph
+    # with choose_qparams
+    if is_dynamic:
```

Comment:
Can be done in a future PR, but right now it's not clear what dtypes are supported in what modes without diving into this code. Ideally we would have a helper function or a constant list somewhere since these dtypes are hardcoded already.

Reply:
I actually think just setting `is_dynamic = [torch.quint8, torch.qint8, torch.float16]` here is clearer, so the reader doesn't have to trace through the if else cases above.

Reply:
I remember @vkuzo was trying to make this clearer by removing compute_dtype, and just use dtype + is_dynamic flag directly, hopefully things will be clearer after that, @vkuzo can you share the current status on this refactor?

Reply:
there was still some discussion on ops vs black box backends, I think we just punted on that, that's why I didn't land the PR yet

it's pretty confusing to use `compute_dtype` and `is_dynamic` at the same time in the same function, can we stick to just one of them?  I'd recommend just using `compute_dtype in [torch.quint8, torch.qint8, torch.float16]`, since that's what the current code is using

Reply:
I think right now the condition for is_dynamic is:

```
not (dtype in [torch.quint8, torch.qint8, torch.qint32] and \
            not hasattr(activation_post_process, 'compute_dtype')) and compute_dtype in [torch.quint8, torch.qint8, torch.float16]
```
since "compute_dtype in [torch.quint8, torch.qint8, torch.float16]" in L174 appears in `elif` branch, would it be OK to use this as a condition to replace is_dynamic?

Reply:
I added a TODO to cleanup the condition for first `if`, this is really hard to reason about, maybe it will get better if we have is_dynamic flag

Reply:
For other PRs in this stack seems ok to fix technical debt later since they are not touching production code, but this PR is directly changing production code, in the part of the codebase that has ranked as really hard to work with for our peers.  I think we should be more careful about adding technical debt here.

Can we either make this better in this PR or just move this code out of the prod path until it's easier to understand?  I think by moving faster on this project by not fixing the debt, we'd be slowing down other developers, so it will be a net negative overall.

Reply:
sounds good, updated